### PR TITLE
Phoenix v1/tif ts

### PR DESCRIPTION
**Pull request by jarry-xiao**
- Update TS sdk to support TIF changes
- Refactor event deserialization
- Add new types
- Modify FIFORestingOrder

- [x] Build in clock functionality to filter out expired tombstones from the book
- [ ] Add CLI args to tests to make them more configurable (mainly by network)

Punting on the CLI args

### CODE REVIEW

File: typescript/phoenix-sdk/config.json [hunk 1]
```diff
@@ -67,7 +67,7 @@
       },
       {
         "name": "Wrapped SOL",
-        "symbol": "SOL",
+        "symbol": "wSOL",
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/config.json:70]: we should make this match mainnet (SOL)

File: typescript/phoenix-sdk/src/market.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { Connection, PublicKey } from "@solana/web3.js";
+import { Connection, PublicKey, SYSVAR_CLOCK_PUBKEY } from "@solana/web3.js";
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:1]: is this import used?

File: typescript/phoenix-sdk/src/types/Seat.ts [hunk 3]
```diff
@@ -5,28 +5,28 @@
...
-  market: web3.PublicKey;
-  trader: web3.PublicKey;
-  approvalStatus: beet.bignum;
-  padding: beet.bignum[] /* size: 16 */;
-};
+  discriminant: beet.bignum
+  market: web3.PublicKey
+  trader: web3.PublicKey
+  approvalStatus: beet.bignum
+  padding: beet.bignum[] /* size: 6 */
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/types/Seat.ts:16]: ?

**Review by eugene-chen — COMMENTED** [review]
Add script that also gets some TIF events

**Comment by eugene-chen** [comment]
Is there a way to configure solita to use our preferred formatting (double-quotes + semicolons)? makes these diffs a lot harder to review.

**Comment by jarry-xiao** [comment]
> Is there a way to configure solita to use our preferred formatting (double-quotes + semicolons)? makes these diffs a lot harder to review.

I'm going to make a PR to `phoenix-v1` that runs prettier on the ts client in the generation script

File: typescript/phoenix-sdk/tests/openOrders.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
+import { Connection } from "@solana/web3.js";
+
+import * as Phoenix from "../src";
+
+const getOrderSequenceNumber = (orderId: Phoenix.OrderId): bigint => {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/tests/openOrders.ts]: is this tested?

File: typescript/phoenix-sdk/tests/openOrders.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  const connection = new Connection("http://127.0.0.1:8899");
+  const phoenix = await Phoenix.Client.create(connection, "localhost");
+
+  const market = Array.from(phoenix.markets.values()).find(
+    (market) => market.name === "SOL/USDC"
+  );
+  if (!market) throw new Error("Market not found");
+
+  let traderKey;
+  for (const [trader, traderState] of market.data.traders) {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/tests/openOrders.ts:20]: this is sample code that people will read. add comment that this is finding the first trader that has something on the book

File: typescript/phoenix-sdk/tests/openOrders.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            ),
+            marketAddress.toString()
+          ),
+          timeRemaining
+        );
+      }
+    }
+
+    for (const [orderId, order] of market.data.bids) {
+      if (Phoenix.toNum(order.traderIndex) === traderIndex) {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/tests/openOrders.ts:81]: pull out duplicated code into function

File: typescript/phoenix-sdk/src/client.ts [hunk 7]
```diff
@@ -8,30 +13,85 @@ import {
...
+    marketKeysToData: Array<[PublicKey, AccountInfo<Buffer>]>,
+    tokenConfig: Array<TokenConfig>
+  ): Promise<[Map<string, Market>, Token[]]> {
+    const tokens = [];
+    const markets = new Map();
+    await Promise.all(
+      // For every market:
+      marketKeysToData.map(async ([marketAddress, marketAccount]) => {
+        // Load the market
+        const market = await Market.load({
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:78]: why is this async? we already have the data, no?

File: typescript/phoenix-sdk/src/client.ts [hunk 8]
```diff
@@ -50,75 +110,326 @@ export class Client {
     const config_url =
       "https://raw.githubusercontent.com/Ellipsis-Labs/phoenix-sdk/master/typescript/phoenix-sdk/config.json";
 
-    const config = await axios.get(config_url).then((response) => {
+    const marketConfigs = await axios.get(config_url).then((response) => {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:118]: can this method just call `createWithMarketAddresses`? lots of duplicated code

File: typescript/phoenix-sdk/src/client.ts [hunk 9]
```diff
@@ -50,75 +110,326 @@ export class Client {
...
+  }
+
+  /**
+   * Refreshes the market data and clock
+   *
+   * @param marketAddress The address of the market to refresh
+   *
+   * @returns The refreshed Market
+   */
+  public async refreshMarket(
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:275]: no `refreshAllMarkets`?

File: typescript/phoenix-sdk/src/client.ts [hunk 10]
```diff
@@ -50,75 +110,326 @@ export class Client {
...
+    return getMarketLadder(
+      market.data,
+      this.clock.slot,
+      this.clock.unixTimestamp,
+      levels
+    );
+  }
+
+  /**
+   * Returns the market's ladder of bids and asks  as JS numbers
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:340]: extra space in doc. also, this doc doesn't really explain the difference between this method and above.

File: typescript/phoenix-sdk/src/client.ts [hunk 11]
```diff
@@ -50,75 +110,326 @@ export class Client {
...
+  }
+
+  /**
+   * Returns the expected amount out for a given swap order
+   *
+   * @param marketAddress The `MarketAddress` for the swap market
+   * @param side The side of the order (Bid or Ask)
+   * @param inAmount The amount of the input token
+   *
+   * TODO this should use getMarketLadder and adjust its calculation
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:371]: is this comment outdated?

File: typescript/phoenix-sdk/src/client.ts [hunk 12]
```diff
@@ -154,27 +463,23 @@ export class Client {
...
   public getSeatKey(trader: PublicKey, marketAddress: string): PublicKey {
-    const market = this.markets.find(
-      (m) => m.address.toBase58() === marketAddress
-    );
+    const market = this.markets.get(marketAddress);
     if (!market) throw new Error("Market not found: " + marketAddress);
-    const market_pubkey = new PublicKey(
-      "14CAwu3LiBBk5fcHGdTsFyVxDwvpgFiSfDwgPJxECcE5"
-    );
+    const marketPubkey = new PublicKey(marketAddress);
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:478]: lol ty for fixing

File: typescript/phoenix-sdk/src/client.ts [hunk 13]
```diff
@@ -183,18 +488,16 @@ export class Client {
   }
 
   /**
-   * Get the price in ticks for a given price 
+   * Get the price in ticks for a given price
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:496]: specify the original units, perhaps add an example

File: typescript/phoenix-sdk/src/client.ts [hunk 14]
```diff
@@ -223,40 +524,36 @@ export class Client {
     rawBaseUnits: number,
     marketAddress: string
   ): number {
-    const market = this.markets.find(
-      (m) => m.address.toBase58() === marketAddress
-    );
+    const market = this.markets.get(marketAddress);
     if (!market) throw new Error("Market not found: " + marketAddress);
-    const base_units = rawBaseUnits / market.data.header.rawBaseUnitsPerBaseUnit;
+    const base_units =
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:534]: `baseUnits`

File: typescript/phoenix-sdk/src/client.ts [hunk 15]
```diff
@@ -223,40 +524,36 @@ export class Client {
     rawBaseUnits: number,
     marketAddress: string
   ): number {
-    const market = this.markets.find(
-      (m) => m.address.toBase58() === marketAddress
-    );
+    const market = this.markets.get(marketAddress);
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:532]: function name should specify rounded down

File: typescript/phoenix-sdk/src/market.ts [hunk 16]
```diff
@@ -94,53 +91,51 @@ export class Market {
    * @param marketAddress The `PublicKey` of the market account
    */
   static async load({
-    connection,
     address,
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:94]: this function should not be async

File: typescript/phoenix-sdk/src/market.ts [hunk 17]
```diff
@@ -94,53 +91,51 @@ export class Market {
...
+    buffer,
+    tokenList,
   }: {
-    connection: Connection;
     address: PublicKey;
+    buffer: Buffer;
+    tokenList: TokenConfig[];
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:100]: does the market config not tell us which tokens belong to the market? we should be able to pass in a base token config and a quote token config directly rather than finding them in the token list

Comment (eugene-chen) [reply typescript/phoenix-sdk/src/market.ts:100]: (upstream of here)

File: typescript/phoenix-sdk/src/market.ts [hunk 18]
```diff
@@ -94,53 +91,51 @@ export class Market {
...
+    const quoteKey = marketData.header.baseParams.mintKey.toBase58();
+    const quoteKeyNameBackup =
+      quoteKey.slice(0, 8) + "..." + quoteKey.slice(-8);
 
     const baseToken = new Token({
-      name: baseTokenConfig.name,
-      symbol: baseTokenConfig.symbol,
-      logoUri: baseTokenConfig.logoUri,
+      name:
+        baseTokenConfig !== undefined
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:119]: if token config is undefined we should fail. are there cases where this is expected?

File: typescript/phoenix-sdk/src/market.ts [hunk 19]
```diff
@@ -237,15 +208,60 @@ export class Market {
...
     inAmount: number;
+    slot: beet.bignum;
+    unixTimestamp: beet.bignum;
   }): number {
-    return getMarketExpectedOutAmount({
-      marketData: this.data,
-      side,
-      inAmount,
-    });
+    const numBids = toNum(this.data.header.marketSizeParams.bidsSize);
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/market.ts:219]: why can't this call the other implementation of getExpectedOutAmount? prefer to not duplicate this code

File: typescript/phoenix-sdk/src/utils/market.ts [hunk 20]
```diff
@@ -189,6 +187,7 @@ export function deserializeRedBlackTree<Key, Value>(
   const nodes = tree_nodes[0];
   const freeNodes = tree_nodes[1];
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/market.ts:190]: why is this needed?

File: typescript/phoenix-sdk/src/utils/market.ts [hunk 21]
```diff
@@ -189,6 +187,7 @@ export function deserializeRedBlackTree<Key, Value>(
   const nodes = tree_nodes[0];
   const freeNodes = tree_nodes[1];
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/market.ts:188]: `treeNodes` and `indexMap`

File: typescript/phoenix-sdk/src/utils/market.ts [hunk 22]
```diff
@@ -275,11 +273,26 @@ function deserializeRedBlackTreeNodes<Key, Value>(
...
   marketData: MarketData,
+  slot: beet.bignum,
+  unixTimestamp: beet.bignum,
   levels: number = DEFAULT_LADDER_DEPTH
 ): Ladder {
   const bids: Array<[BN, BN]> = [];
   const asks: Array<[BN, BN]> = [];
   for (const [orderId, restingOrder] of marketData.bids) {
+    if (restingOrder.lastValidSlot != 0) {
+      if (restingOrder.lastValidSlot < slot) {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/market.ts:284]: please combine these into a single if

File: typescript/phoenix-sdk/src/utils/market.ts [hunk 23]
```diff
@@ -275,11 +273,26 @@ function deserializeRedBlackTreeNodes<Key, Value>(
...
+        continue;
+      }
+    }
+    if (restingOrder.lastValidUnixTimestampInSeconds != 0) {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/market.ts:288]: here too

File: typescript/phoenix-sdk/src/utils/market.ts [hunk 24]
```diff
@@ -301,6 +314,19 @@ export function getMarketLadder(
   }
 
   for (const [orderId, restingOrder] of marketData.asks) {
+    if (restingOrder.lastValidSlot != 0) {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/market.ts:317]: here too

File: typescript/phoenix-sdk/src/utils/market.ts [hunk 25]
```diff
@@ -301,6 +314,19 @@ export function getMarketLadder(
...
+      if (restingOrder.lastValidSlot < slot) {
+        continue;
+      }
+    }
+    if (restingOrder.lastValidUnixTimestampInSeconds != 0) {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/market.ts:322]: here too

File: typescript/phoenix-sdk/tests/deserializeClock.ts [hunk 26]
```diff
@@ -0,0 +1,28 @@
+import { Connection } from "@solana/web3.js";
+import { BN } from "bn.js";
+
+import * as Phoenix from "../src";
+
+// Ex: ts-node tests/market.ts
+export async function market() {
+
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/tests/deserializeClock.ts:8]: indent here is not consistent with rest of repo

File: typescript/phoenix-sdk/tests/watch.ts [hunk 27]
```diff
@@ -3,25 +3,30 @@ import { Connection } from "@solana/web3.js";
...
       console.clear();
       console.log("Ladder update", updates + 1, "of", 10, "\n");
-      market.printLadder();
+      phoenix.printLadder(marketAddress);
       lastLadder = ladder;
       updates++;
     }
 
-    await market.refresh(connection);
+    await phoenix.refreshMarket(marketAddress);
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/tests/watch.ts:29]: not sure how i feel about this interface. you can have a clock out of sync with a different market

File: typescript/phoenix-sdk/src/client.ts [hunk 28]
```diff
@@ -86,7 +86,12 @@ export class Client {
         // Set the tokens from the market (avoiding duplicates)
         for (const token of [market.baseToken, market.quoteToken]) {
           const mint = token.data.mintKey.toBase58();
-          if (tokens.find((t) => t.data.mintKey.toBase58() === mint)) continue;
+          if (
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:89]: `tokens` should probably be an Object keyed by token mint as string

File: typescript/phoenix-sdk/src/market.ts (hunk 18 above)
Comment (jarry-xiao) [reply typescript/phoenix-sdk/src/market.ts:119]: I think it's totally fine to use this with a market without a defined config! Don't see a reason why the API should be that restrictive

File: typescript/phoenix-sdk/src/utils/market.ts (hunk 20 above)
Comment (jarry-xiao) [reply typescript/phoenix-sdk/src/utils/market.ts:190]: It treats the underscore as a used variable for some reason....

Comment (eugene-chen) [reply typescript/phoenix-sdk/src/utils/market.ts:190]: lol. I think you can do
`for (const [index, [key, ]] of node.entries())`

File: typescript/phoenix-sdk/src/fetchDecodeInstructions.ts [hunk 29]
```diff
@@ -1,147 +1,238 @@
...
-
-    }
-
-    return JSON.stringify({ instructionName, accounts, decodedData });
-
+  const decoded = bs58.decode(data.toString());
+
+  const instructionEnum = decoded[0];
+
+  const matched_idl_instruction = idl.instructions.filter((instruction) => {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/fetchDecodeInstructions.ts:34]: `matchedIdlInstruction`

**Comment by eugene-chen** [comment]
Last thing I want to see is an example that touches multiple markets (probably devnet since it already has multiple markets).

Get multiple books, refresh one or two books, send an order to one
